nltk/translate/nist_score.py

```diff
@@ -93,41 +93,46 @@ def corpus_nist(list_of_references, hypotheses, n=5):
     # Before proceeding to compute NIST, perform sanity checks.
     assert len(list_of_references) == len(hypotheses), "The number of hypotheses and their reference(s) should be the same"
 
-    p_numerators = Counter() # Key = ngram order, and value = no. of ngram matches.
-    p_denominators = Counter() # Key = ngram order, and value = no. of ngram in ref.
-    sysoutput_lengths = Counter() # Key = ngram order, and value = no. of ngram in hyp.
-    hyp_lengths, ref_lengths = 0, 0
+    # Compute the information weights based on the reference sentences.
+    ngram_freq = Counter()
+    for references in list_of_references: # For each source sent, there's a list of reference sents.
+        for reference in references:
+            # For each order of ngram, count the ngram occurrences.
+            for i, _ in enumerate(range(1,n+1)):
+                ngram_freq.update(ngrams(reference, i))
 
+    hyp_lengths, ref_lengths = 0, 0
+    sysoutput_lengths = Counter()    # Key = ngram order, and value = no. of ngram in hyp.
+    information_weights = Counter()  # Key = ngram order, and value = sum of info weights per nth order.
     # Iterate through each hypothesis and their corresponding references.
     for references, hypothesis in zip(list_of_references, hypotheses):
-        # For each order of ngram, calculate the numerator and
-        # denominator for the corpus-level modified precision.
-        for i, _ in enumerate(range(1,n+1)):
-            p_i = modified_precision(references, hypothesis, i)
-            p_numerators[i] += p_i.numerator
-            p_denominators[i] += p_i.denominator
-            # Adds the no. of ngrams in the hypothesis.
-            sysoutput_lengths[i] += len(hypothesis) - (i - 1)
-
         # Calculate the hypothesis length and the closest reference length.
         # Adds them to the corpus-level hypothesis and reference counts.
         hyp_len =  len(hypothesis)
         hyp_lengths += hyp_len
         ref_lengths += closest_ref_length(references, hyp_len)
 
+        # For each order of ngram.
+        for i, _ in enumerate(range(1,n+1)):
+            # Counter of ngrams in hypothesis.
+            hyp_ngrams = Counter(ngrams(hypothesis, i)) if len(hypothesis) >= i else Counter()
+            # Adds the no. of ngrams in the hypothesis.
+            sysoutput_lengths[i] += len(hyp_ngrams)
+            # Compute the information weights per overlapped ngram.
+            for ng in hyp_ngrams:
+                # Eqn 2 in Doddington (2002):
+                # Info(w_1 ... w_n) = log_2 [ (# of occurrences of w_1 ... w_n-1) / (# of occurrences of w_1 ... w_n) ]
+                numerator = ngram_freq[ng]
+                denominator = ngram_freq[ng[:-1]] if i > 1 else hyp_len
+                if numerator == 0 or denominator == 0:
+                    information_weights[i] += 0
+                else:
+                    information_weights[i] += -1 * math.log(numerator/ denominator, 2)
```

Comment:
this should have `-hyp_ngrams[ng]` instead of -1. In fact, not `-hyp_ngrams[ng]` but `-min(hyp_ngrams[ng], max(number of times ng occurs in reference_text for reference_text in reference_texts))`

Reply:
Doubling checking my understanding.

Since we're going to need ` max(number of times ng occurs in reference_text for reference_text in reference_texts))`, we can pre-compute these before going through the references while counting the ngrams in the reference texts, i.e. `max_ngram_freq`:

```
# Compute the information weights based on the reference sentences.
    ngram_freq = Counter()
    max_ngram_freq = Counter()
    for references in list_of_references: # For each source sent, there's a list of reference sents.
        for reference in references:
            # For each order of ngram, count the ngram occurrences.
            for i, _ in enumerate(range(1,n+1)):
                ng_i = ngrams(reference, i)
                ngram_freq.update(ng_i)
                for ng, count in Counter(ng_i).items():
                    if count > max_ngram_freq[ng]:
                        max_ngram_freq[ng] = count
```

Reply:
And to compute the `information_weights[i]`, it would be:

```
if numerator == 0 or denominator == 0:
    information_weights[i] += 0
else:
    information_weights[i] += numerator/ denominator
```

And the `-log2()` can be done at the final return sum?

```
return sum(-math.log2(info_i)/sysoutput_lengths[i] for i, info_i in information_weights.items()) * bp
```